File: src/lib/setup/bracketing.ts
import type { BracketingRun, BracketingSeries } from "@/types/garage";

/**
 * F-SET-003 Bracketing-Auswertung
 * Effekt belegt nur wenn |Δ| > 1,5 × gepoolte SD und n ≥ 2 je Konfiguration.
 */

export interface ConfigStats {
  value: number;
  n: number;
  meanTime: number;
  meanFlow: number;
  meanImpact: number;
  meanSubjective: number;
  sdTime: number;
  sdFlow: number;
}

function mean(xs: number[]): number {
  if (xs.length === 0) return 0;
  return xs.reduce((a, b) => a + b, 0) / xs.length;
}

function sampleSd(xs: number[]): number {
  if (xs.length < 2) return 0;
  const m = mean(xs);
  const v = xs.reduce((s, x) => s + (x - m) ** 2, 0) / (xs.length - 1);
  return Math.sqrt(v);
}

export function groupRunsByConfig(runs: BracketingRun[]): Map<number, BracketingRun[]> {
  const map = new Map<number, BracketingRun[]>();
  for (const r of runs) {
    const list = map.get(r.configValue) ?? [];
    list.push(r);
    map.set(r.configValue, list);
  }
  return map;
}

export function statsForConfig(value: number, runs: BracketingRun[]): ConfigStats {
  const times = runs.map((r) => r.segmentTimeSec);
  const flows = runs.map((r) => r.flowScore);
  const impacts = runs.map((r) => r.impactHardness);
  const subs = runs.map((r) => r.subjectiveRating);
  return {
    value,
    n: runs.length,
    meanTime: mean(times),
    meanFlow: mean(flows),
    meanImpact: mean(impacts),
    meanSubjective: mean(subs),
    sdTime: sampleSd(times),
    sdFlow: sampleSd(flows),
  };
}

function pooledSd(a: ConfigStats, b: ConfigStats, key: "sdTime" | "sdFlow"): number {
  const n1 = a.n;
  const n2 = b.n;
  if (n1 + n2 <= 2) return 0;
  const s1 = a[key];
  const s2 = b[key];
  return Math.sqrt(((n1 - 1) * s1 ** 2 + (n2 - 1) * s2 ** 2) / (n1 + n2 - 2));
}

export interface BracketingEvaluation {
  noProvenDifference: boolean;
  provenBestValue?: number;
  summary: string;
  ready: boolean;
  missingRuns: { value: number; have: number; need: number }[];
  comparisons: {
    a: number;
    b: number;
    metric: string;
    delta: number;
    threshold: number;
    proven: boolean;
  }[];
}

/** Composite-Score: höher = besser (Flow↑, Zeit↓, Impact↓, Subjective↑) */
function composite(s: ConfigStats): number {
  return (
    s.meanFlow * 0.35 +
    s.meanSubjective * 20 * 0.25 +
    (1 / Math.max(1, s.meanTime)) * 100 * 0.2 +
    (10 - Math.min(10, s.meanImpact)) * 10 * 0.2
  );
}

export function evaluateBracketingSeries(series: BracketingSeries): BracketingEvaluation {
  const grouped = groupRunsByConfig(series.runs);
  const values = Array.from(grouped.keys()).sort((a, b) => a - b);
  const missingRuns: BracketingEvaluation["missingRuns"] = [];

  // Erwartete Konfigurationen aus Range
  const expected: number[] = [];
  for (let v = series.rangeFrom; v <= series.rangeTo + 1e-9; v += series.step) {
    expected.push(Math.round(v * 1000) / 1000);
  }

  for (const v of expected) {
    const have = grouped.get(v)?.length ?? 0;
    if (have < 2) missingRuns.push({ value: v, have, need: 2 });
  }

  if (missingRuns.length > 0 || values.length < 2) {
    return {
      noProvenDifference: false,
      summary:
        missingRuns.length > 0
          ? `Noch nicht auswertbar: pro Konfiguration mind. 2 Durchgänge nötig. Fehlt: ${missingRuns
              .map((m) => `${m.value} (${m.have}/2)`)
              .join(", ")}`
          : "Mindestens zwei verschiedene Konfigurationen mit je ≥ 2 Runs nötig.",
      ready: false,
      missingRuns,
      comparisons: [],
    };
  }

  const stats = values.map((v) => statsForConfig(v, grouped.get(v)!));
  const comparisons: BracketingEvaluation["comparisons"] = [];
  let anyProven = false;

  for (let i = 0; i < stats.length; i++) {
    for (let j = i + 1; j < stats.length; j++) {
      const a = stats[i];
      const b = stats[j];
      // Spec: |Δ| > 1,5× gepoolte SD. Zusätzlich Mindest-Effektgröße,
      // damit numerisches Rauschen bei SD≈0 keinen „Beweis“ erzeugt (Blindtest).
      const sdFlow = pooledSd(a, b, "sdFlow");
      const deltaFlow = Math.abs(a.meanFlow - b.meanFlow);
      const thrFlow = Math.max(1.5 * sdFlow, 1.0);
      const provenFlow = a.n >= 2 && b.n >= 2 && deltaFlow > thrFlow;

      const sdTime = pooledSd(a, b, "sdTime");
      const deltaTime = Math.abs(a.meanTime - b.meanTime);
      const thrTime = Math.max(1.5 * sdTime, 0.5);
      const provenTime = a.n >= 2 && b.n >= 2 && deltaTime > thrTime;

      comparisons.push({
        a: a.value,
        b: b.value,
        metric: "flow",
        delta: deltaFlow,
        threshold: thrFlow,
        proven: provenFlow,
      });
      comparisons.push({
        a: a.value,
        b: b.value,
        metric: "time",
        delta: deltaTime,
        threshold: thrTime,
        proven: provenTime,
      });
      if (provenFlow || provenTime) anyProven = true;
    }
  }

  if (!anyProven) {
    return {
      noProvenDifference: true,
      summary:
        "Kein belegbarer Unterschied — die Lauf-zu-Lauf-Streuung übersteigt die gemessenen Differenzen (|Δ| ≤ 1,5× gepoolte SD). Das ist ein gültiges Ergebnis (F-SET-003).",
      ready: true,
      missingRuns: [],
      comparisons,
    };
  }

  const ranked = [...stats].sort((a, b) => composite(b) - composite(a));
  const best = ranked[0];

  return {
    noProvenDifference: false,
    provenBestValue: best.value,
    summary: `Belegbare Unterschiede vorhanden. Beste Konfiguration für ${series.parameter}: ${best.value} ${series.unit} (höherer Flow / bessere Segmentzeit bei ausreichender Effektstärke).`,
    ready: true,
    missingRuns: [],
    comparisons,
  };
}

/** Blindtest-Helfer: zwei identische Configs → sollte „kein Unterschied“ ergeben */
export function blindTestIdenticalSetups(
  runsA: BracketingRun[],
  runsB: BracketingRun[]
): boolean {
  const series: BracketingSeries = {
    id: "blind",
    bikeId: "x",
    setupId: "x",
    parameter: "fork.rebound",
    unit: "clicks",
    rangeFrom: 6,
    rangeTo: 8,
    step: 2,
    referenceSegmentLabel: "test",
    status: "open",
    runs: [
      ...runsA.map((r) => ({ ...r, configValue: 6 })),
      ...runsB.map((r) => ({ ...r, configValue: 8 })),
    ],
    createdAt: new Date().toISOString(),
  };
  const result = evaluateBracketingSeries(series);
  return result.ready && !!result.noProvenDifference;
}

File: mobile/lib/domain/setup/bracketing.dart
import 'dart:math' as math;

import '../ride.dart';

/// F-SET-003 Bracketing-Auswertung (Port src/lib/setup/bracketing.ts).

class BracketingRun {
  const BracketingRun({
    required this.configValue,
    required this.segmentTimeSec,
    required this.flowScore,
    required this.impactHardness,
    required this.subjectiveRating,
  });

  final double configValue;
  final double segmentTimeSec;
  final double flowScore;
  final double impactHardness;
  final double subjectiveRating;
}

class BracketingSeries {
  const BracketingSeries({
    required this.adjusterKey,
    required this.rangeFrom,
    required this.rangeTo,
    required this.step,
    required this.runs,
  });

  final String adjusterKey;
  final double rangeFrom;
  final double rangeTo;
  final double step;
  final List<BracketingRun> runs;
}

class ConfigStats {
  const ConfigStats({
    required this.value,
    required this.n,
    required this.meanTime,
    required this.meanFlow,
    required this.meanImpact,
    required this.meanSubjective,
    required this.sdTime,
    required this.sdFlow,
  });

  final double value;
  final int n;
  final double meanTime;
  final double meanFlow;
  final double meanImpact;
  final double meanSubjective;
  final double sdTime;
  final double sdFlow;
}

class BracketingEvaluation {
  const BracketingEvaluation({
    required this.noProvenDifference,
    required this.summary,
    required this.ready,
    required this.missingRuns,
    this.provenBestValue,
  });

  final bool noProvenDifference;
  final double? provenBestValue;
  final String summary;
  final bool ready;
  final List<({double value, int have, int need})> missingRuns;
}

double _mean(List<double> xs) {
  if (xs.isEmpty) return 0;
  return xs.reduce((a, b) => a + b) / xs.length;
}

double _sd(List<double> xs) {
  if (xs.length < 2) return 0;
  final m = _mean(xs);
  final v =
      xs.map((x) => (x - m) * (x - m)).reduce((a, b) => a + b) / (xs.length - 1);
  return math.sqrt(v);
}

ConfigStats statsForConfig(double value, List<BracketingRun> runs) {
  return ConfigStats(
    value: value,
    n: runs.length,
    meanTime: _mean(runs.map((r) => r.segmentTimeSec).toList()),
    meanFlow: _mean(runs.map((r) => r.flowScore).toList()),
    meanImpact: _mean(runs.map((r) => r.impactHardness).toList()),
    meanSubjective: _mean(runs.map((r) => r.subjectiveRating).toList()),
    sdTime: _sd(runs.map((r) => r.segmentTimeSec).toList()),
    sdFlow: _sd(runs.map((r) => r.flowScore).toList()),
  );
}

double _composite(ConfigStats s) {
  return s.meanFlow * 0.35 +
      s.meanSubjective * 20 * 0.25 +
      (1 / math.max(1, s.meanTime)) * 100 * 0.2 +
      (10 - math.min(10, s.meanImpact)) * 10 * 0.2;
}

double _pooledSdFlow(ConfigStats a, ConfigStats b) {
  final n1 = a.n;
  final n2 = b.n;
  if (n1 + n2 <= 2) return 0;
  return math.sqrt(
    ((n1 - 1) * a.sdFlow * a.sdFlow + (n2 - 1) * b.sdFlow * b.sdFlow) /
        (n1 + n2 - 2),
  );
}

/// MVP: zwei Varianten ±step um aktuellen Wert (Blind-Hinweis in UI).
BracketingSeries createBlindPair({
  required String adjusterKey,
  required double currentValue,
  double step = 2,
  List<BracketingRun> runs = const [],
}) {
  return BracketingSeries(
    adjusterKey: adjusterKey,
    rangeFrom: currentValue - step,
    rangeTo: currentValue + step,
    step: step * 2,
    runs: runs,
  );
}

/// Baut Bracketing-Runs aus gespeicherten Rides (Feedback + Summary).
/// Ordnet Runs abwechselnd A/B zu, wenn genug Rides mit Feedback vorhanden.
List<BracketingRun> runsFromRides({
  required List<RideRecord> rides,
  required double configA,
  required double configB,
  String? bikeId,
}) {
  final filtered = [
    for (final r in rides)
      if (bikeId == null || r.bikeId == bikeId)
        if (r.feedback != null && r.feedback!.skipped != true) r,
  ];
  if (filtered.isEmpty) return const [];

    final runs = <BracketingRun>[];
  for (var i = 0; i < filtered.length; i++) {
    final r = filtered[i];
    final cfg = i.isEven ? configA : configB;
    final flow = (r.summary['avgFlow'] as num?)?.toDouble() ?? 60;
    final feel = (r.feedback?.overallFeel ?? 3).toDouble();
    final harsh = r.feedback?.smallBump == 'harsh'
        ? 6.0
        : r.feedback?.frontFeel == 'too_firm'
            ? 5.0
            : 3.5;
    runs.add(
      BracketingRun(
        configValue: cfg,
        segmentTimeSec: (r.movingTimeSec > 0 ? r.movingTimeSec : 600).toDouble(),
        flowScore: flow.clamp(0, 100),
        impactHardness: harsh.clamp(1, 10),
        subjectiveRating: feel.clamp(1, 5),
      ),
    );
  }
  return runs;
}

/// Demo-Runs nur als Fallback wenn keine Ride-Daten.
List<BracketingRun> syntheticBracketRuns(double a, double b) => [
      BracketingRun(
        configValue: a,
        segmentTimeSec: 120,
        flowScore: 70,
        impactHardness: 4,
        subjectiveRating: 3,
      ),
      BracketingRun(
        configValue: a,
        segmentTimeSec: 118,
        flowScore: 72,
        impactHardness: 3.5,
        subjectiveRating: 4,
      ),
      BracketingRun(
        configValue: b,
        segmentTimeSec: 125,
        flowScore: 68,
        impactHardness: 5,
        subjectiveRating: 3,
      ),
      BracketingRun(
        configValue: b,
        segmentTimeSec: 122,
        flowScore: 69,
        impactHardness: 4.5,
        subjectiveRating: 3,
      ),
    ];

BracketingEvaluation evaluateBracketingSeries(BracketingSeries series) {
  final grouped = <double, List<BracketingRun>>{};
  for (final r in series.runs) {
    grouped.putIfAbsent(r.configValue, () => []).add(r);
  }
  final missing = <({double value, int have, int need})>[];
  final expected = <double>[];
  for (var v = series.rangeFrom; v <= series.rangeTo + 1e-9; v += series.step) {
    expected.add((v * 1000).round() / 1000);
  }
  for (final v in expected) {
    final have = grouped[v]?.length ?? 0;
    if (have < 2) missing.add((value: v, have: have, need: 2));
  }

  if (missing.isNotEmpty || grouped.keys.length < 2) {
    return BracketingEvaluation(
      noProvenDifference: false,
      ready: false,
      missingRuns: missing,
      summary: missing.isNotEmpty
          ? 'Noch nicht auswertbar: je Konfiguration ≥ 2 Runs. Fehlt: ${missing.map((m) => '${m.value} (${m.have}/2)').join(', ')}'
          : 'Mindestens zwei Konfigurationen mit je ≥ 2 Runs nötig.',
    );
  }

  final stats = [
    for (final e in grouped.entries) statsForConfig(e.key, e.value),
  ]..sort((a, b) => a.value.compareTo(b.value));

  var provenBest = stats.first;
  for (final s in stats.skip(1)) {
    if (_composite(s) > _composite(provenBest)) provenBest = s;
  }

  var anyProven = false;
  for (var i = 0; i < stats.length; i++) {
    for (var j = i + 1; j < stats.length; j++) {
      final pooled = _pooledSdFlow(stats[i], stats[j]);
      final delta = (stats[i].meanFlow - stats[j].meanFlow).abs();
      if (pooled > 0 && delta > 1.5 * pooled) anyProven = true;
    }
  }

  if (!anyProven) {
    return const BracketingEvaluation(
      noProvenDifference: true,
      ready: true,
      missingRuns: [],
      summary:
          'Kein belegbarer Unterschied (Δ ≤ 1,5× SD). Blind-Test: Setups wirkten praktisch gleich.',
    );
  }

  return BracketingEvaluation(
    noProvenDifference: false,
    provenBestValue: provenBest.value,
    ready: true,
    missingRuns: const [],
    summary:
        'Belegbarer Vorteil bei ${series.adjusterKey} = ${provenBest.value.toStringAsFixed(0)} (Composite-Score).',
  );
}
